src/Api/SecretsManager/Controllers/SecretsController.cs

```diff
@@ -18,62 +22,196 @@ public class SecretsController : Controller
-        return new SecretWithProjectsListResponseModel(secrets);
+        var secretsAllowedToRead = new List<Secret>();
+
+        foreach(Secret secret in secrets)
+        {
+           if(await userHasReadAccessToProject(secret))
+           {
+               secretsAllowedToRead.Add(secret);
+           }
+        }
+
+        return new SecretWithProjectsListResponseModel(secretsAllowedToRead);
     }
 
     [HttpGet("secrets/{id}")]
     public async Task<SecretResponseModel> GetSecretAsync([FromRoute] Guid id)
     {
         var secret = await _secretRepository.GetByIdAsync(id);
-        if (secret == null)
+        if(await userHasReadAccessToProject(secret))
+        {
+            if (secret == null)
+            {
+                throw new NotFoundException();
+            }
+            return new SecretResponseModel(secret);
+        } 
+        else 
         {
-            throw new NotFoundException();
+            throw new UnauthorizedAccessException("You don't have read access");
         }
-        return new SecretResponseModel(secret);
+
+        return null;
     }
 
     [HttpGet("projects/{projectId}/secrets")]
     public async Task<SecretWithProjectsListResponseModel> GetSecretsByProjectAsync([FromRoute] Guid projectId)
     {
         var secrets = await _secretRepository.GetManyByProjectIdAsync(projectId);
-        var responses = secrets.Select(s => new SecretResponseModel(s));
-        return new SecretWithProjectsListResponseModel(secrets);
+        
+        if(secrets != null){
+            var firstSecret = secrets.FirstOrDefault();
+            if(await userHasReadAccessToProject(projectId, firstSecret.OrganizationId))
+            {
+                var responses = secrets.Select(s => new SecretResponseModel(s));
+                return new SecretWithProjectsListResponseModel(secrets);
+            } 
+            else
+            {
+                throw new UnauthorizedAccessException("You don't have read access");
+            }
+        }
+
+        return null;
     }
 
     [HttpPost("organizations/{organizationId}/secrets")]
     public async Task<SecretResponseModel> CreateSecretAsync([FromRoute] Guid organizationId, [FromBody] SecretCreateRequestModel createRequest)
     {
-        var result = await _createSecretCommand.CreateAsync(createRequest.ToSecret(organizationId));
-        return new SecretResponseModel(result);
+        var userId = _userService.GetProperUserId(User).Value;
+        var secret = createRequest.ToSecret(organizationId);
+        if(await userHasWriteAccessToProject(secret))
+        {
+            var result = await _createSecretCommand.CreateAsync(createRequest.ToSecret(organizationId), userId);
+            return new SecretResponseModel(result);
+        }
+        else
+        {
+            throw new UnauthorizedAccessException("You don't have read access");
+        }
+
+        return null;
```

Comment:
```
var userId = _userService.GetProperUserId(User).Value;
var secret = createRequest.ToSecret(organizationId);

if(!await userHasWriteAccessToProject(secret))
{
     throw new UnauthorizedAccessException();
}

var result = await _createSecretCommand.CreateAsync(createRequest.ToSecret(organizationId), userId);
return new SecretResponseModel(result);
```

The `return null;` is confusing, and the error message had `Youd don't have read access` when we were checking write access.

Reply:
This is perfect, thank you.